mlflow/tracking/multimedia.py

```diff
@@ -0,0 +1,176 @@
+"""
+Internal module implementing multi-media objects and utilities in MLflow. Multi-media objects are
+exposed to users at the top-level :py:mod:`mlflow` module.
+"""
+
+from typing import TYPE_CHECKING, Optional, Union
+
+if TYPE_CHECKING:
+    import numpy
+    import PIL
+
+COMPRESSED_IMAGE_SIZE = 256
+
+
+def _compress_image_size(
+    image: "PIL.Image.Image", max_size: Optional[int] = COMPRESSED_IMAGE_SIZE
+) -> "PIL.Image.Image":
+    """
+    Scale the image to fit within a square with length `max_size` while maintaining
+    the aspect ratio.
+    """
+    # scale the image to max(width, height) <= compressed_file_max_size
+    width, height = image.size
+    if width > height:
+        new_width = max_size
+        new_height = int(height * (new_width / width))
+    else:
+        new_height = max_size
+        new_width = int(width * (new_height / height))
+    return image.resize((new_width, new_height))
+
+
+def _convert_numpy_to_pil_image(image: Union["numpy.ndarray", list]) -> "PIL.Image.Image":
+    """
+    Convert a numpy array to a PIL image.
+    """
+    import numpy as np
+
+    try:
+        from PIL import Image
+    except ImportError as exc:
+        raise ImportError(
+            "Pillow is required to serialize a numpy array as an image. "
+            "Please install it via: pip install Pillow"
+        ) from exc
+
+    def _normalize_to_uint8(x):
+        is_int = np.issubdtype(x.dtype, np.integer)
+        low = 0
+        high = 255 if is_int else 1
+        if x.min() < low or x.max() > high:
+            if is_int:
+                raise ValueError(
+                    "Integer pixel values out of acceptable range [0, 255]. "
+                    f"Found minimum value {x.min()} and maximum value {x.max()}. "
+                    "Ensure all pixel values are within the specified range."
+                )
+            else:
+                raise ValueError(
+                    "Float pixel values out of acceptable range [0.0, 1.0]. "
+                    f"Found minimum value {x.min()} and maximum value {x.max()}. "
+                    "Ensure all pixel values are within the specified range."
+                )
+
+        # float or bool
+        if not is_int:
+            x = x * 255
+
+        return x.astype(np.uint8)
+
+    # Ref.: https://numpy.org/doc/stable/reference/generated/numpy.dtype.kind.html#numpy-dtype-kind
+    valid_data_types = {
+        "b": "bool",
+        "i": "signed integer",
+        "u": "unsigned integer",
+        "f": "floating",
+    }
+
+    if image.dtype.kind not in valid_data_types:
+        raise TypeError(
+            f"Invalid array data type: '{image.dtype}'. "
+            f"Must be one of {list(valid_data_types.values())}"
+        )
+
+    if image.ndim not in [2, 3]:
+        raise ValueError(f"`image` must be a 2D or 3D array but got a {image.ndim}D array")
```

Comment:
maybe we should just print out the image shape here? As a user, I will like to see what's my image shape, e.g, `[256, 256, 3, 3]` if it's wrong.